### DISCUSSION TITLE
error-stack: how is `Context` stored?

### DISCUSSION THREAD

**Discussion by Abhicodes-crypto** [hashintel/hash#1343] [error-stack (Help)] [answered]
In the current_context function I see that C: `static lifetime. Does that mean C is stored for the entire runtime of the process.

PS: I'm new to rust

**Comment by TimDiekmann** [comment, marked as answer]
Hi @Abhicodes-crypto and thanks for the question!

`C` internally is stored as [`Box<dyn Any>`](https://doc.rust-lang.org/std/any/trait.Any.html) inside of `Report`. This means that `C` is stored on the heap and **will be deallocated as soon as `Report` goes out of scope**.

A little bit of context; `C` requires `'static` because of two reasons:
- If `C` wouldn't be `'static`, the lifetime would bubble up to `Report`, so it would become something like `Report<'a, C>`. As `Report` typically is used as a return value from a function, dealing with lifetimes is hard, the main issue, however, is that
- the exact type of `C` is not known anymore at compile time. To allow downcasting `C`, it's required, that `C` is `'static` as `Any` is implemented for any `T` where `T: 'static`. If I remember correctly there are a few attempts to allow `Any` to have any lifetime but this is a bit ... tricky.

I hope this can answer your question 🙂
